kv_connectors/llmd_fs_backend/src/csrc/storage/file_io.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2025 The llm-d Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <filesystem>
+#include <fstream>
+#include <vector>
+#include <cstring>
+#include <cerrno>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <random>
+
+#include "file_io.hpp"
+#include "thread_pool.hpp"
+
+namespace fs = std::filesystem;
+
+// -------------------------------------------------------------------
+// Constants and thread-local buffers
+// -------------------------------------------------------------------
+// Define a larger buffer (1MB) to reduce syscall overhead and speed up I/O
+const size_t WRITE_BUFFER_SIZE = 1 * 1024 * 1024;  // 1MB buffer
+// Allocate custom I/O buffer for this thread (replaces small default buffer)
+thread_local std::vector<char> thread_write_buffer(WRITE_BUFFER_SIZE);
+// Thread-local unique id for temporary files
+thread_local uint64_t thread_unique_id = std::random_device{}();
+// -------------------------------------------------------------------
+// file-IO Functions
+// -------------------------------------------------------------------
+// Write a tensor to disk using a temporary file and atomic rename
+bool write_tensor_to_file(const torch::Tensor& cpu_tensor,
+                          const std::string& target_path) {
+  TORCH_CHECK(cpu_tensor.is_cpu(), "Tensor must be on CPU");
+  TORCH_CHECK(cpu_tensor.is_contiguous(), "Tensor must be contiguous");
+  // Pointer and size of data to write
+  const void* data_ptr = cpu_tensor.data_ptr();
+  size_t nbytes = cpu_tensor.nbytes();
+
+  // Create parent directory if needed
+  fs::path file_path(target_path);
+  fs::path parent_dir = file_path.parent_path();
+  try {
+    fs::create_directories(parent_dir);
+  } catch (const fs::filesystem_error& e) {
+    std::cerr << "[ERROR] Failed to create directories: " << e.what() << "\n";
+    return false;
+  }
+
+  // Write to a temporary file to ensure atomic replace on rename
+  // Include thread_unique_id so each thread uses a unique temporary file
+  std::string tmp_path =
+      target_path + std::to_string(thread_unique_id) + ".tmp.";
+
+  std::ofstream ofs(tmp_path, std::ios::out | std::ios::binary);
+  if (!ofs) {
+    std::cerr << "[ERROR] Failed to open temporary file for writing: "
```

Comment:
can we add a log_error, log_info, etc. kind util functions that will automatically add the `[LEVEL] ` prefix?